### PR TITLE
Split system and introspection bootstrap sizes

### PR DESCRIPTION
**Pull request by jubrad**
introduces two new flags
         - bootstrap_builtin_system_cluster_size
         - bootstrap_builtin_introspection_cluster_size

        removes the generic builtin cluster size flag
         - bootstrap_builtin_cluster_size

### Motivation
  * This PR is part of a known-desirable feature.
  https://github.com/MaterializeInc/materialize/issues/24983

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/catalog/src/config.rs [hunk 1]
```diff
@@ -54,8 +54,10 @@ pub struct StateConfig {
     pub skip_migrations: bool,
     /// Map of strings to corresponding compute replica sizes.
     pub cluster_replica_sizes: ClusterReplicaSizeMap,
-    /// Builtin cluster replica size.
-    pub builtin_cluster_replica_size: String,
+    /// Builtin system cluster replica size.
+    pub builtin_system_cluster_replica_size: String,
+    /// Builtin intorpsection cluster replica size.
```
Comment (jkosh44) [inline src/catalog/src/config.rs:59]:
```suggestion
    /// Builtin introspection cluster replica size.
```

File: src/environmentd/src/test_util.rs [hunk 2]
```diff
@@ -223,11 +225,19 @@ impl TestHarness {
...
+    pub fn with_builtin_system_cluster_replica_size(
         mut self,
-        builtin_cluster_replica_size: String,
+        builtin_system_cluster_replica_size: String,
     ) -> Self {
-        self.builtin_cluster_replica_size = builtin_cluster_replica_size;
+        self.builtin_system_cluster_replica_size = builtin_system_cluster_replica_size;
+        self
+    }
+    pub fn with_builtin_introspectino_cluster_replica_size(
```
Comment (jkosh44) [inline src/environmentd/src/test_util.rs:235]:
```suggestion
    pub fn with_builtin_introspection_cluster_replica_size(
```

**Review by jkosh44 — APPROVED** [review]
LGTM!

I triggered a Nightly build in case any of the nightly tests were using the removed CLI arg: https://buildkite.com/materialize/nightlies/builds/6955

File: src/adapter/src/catalog/open.rs [hunk 3]
```diff
@@ -1766,15 +1787,18 @@ fn add_new_builtin_cluster_replicas_migration(
 }
 
 pub(crate) fn builtin_cluster_replica_config(
-    builtin_cluster_replica_size: String,
+    replica_size: (String, Option<&ReplicaAllocation>),
```
Comment (jkosh44) [inline src/adapter/src/catalog/open.rs:1790]: Let's just make this two separate arguments.

File: src/adapter/src/catalog/open.rs [hunk 4]
```diff
@@ -1766,15 +1787,18 @@ fn add_new_builtin_cluster_replicas_migration(
...
             availability_zone: None,
             billed_as: None,
-            disk: false,
+            disk: replica_size
+                .1
+                .map(|r| r.disk_limit.is_some())
+                .unwrap_or(false),
             internal: false,
-            size: builtin_cluster_replica_size,
+            size: replica_size.0,
```
Comment (jkosh44) [inline src/adapter/src/catalog/open.rs:1801]: Is the rest of these hard-coded values correct?

Comment (jubrad) [reply src/adapter/src/catalog/open.rs:1801]: uh... yeah, I think we can take the rest of these values, but TBH I don't know what happens if someone were to create a replica that replaces the builtin one. I assume we wouldn't inherit these values, although I'm assuming we'd have to take special care to ensure it's un-billed.

Comment (jkosh44) [reply src/adapter/src/catalog/open.rs:1801]: This is only used during startup, so if we create a replica on a running instance, then it's configuration is determined from the DDL command itself.

**Review by jkosh44 — APPROVED** [review]
LGTM

File: src/adapter/src/catalog/open.rs (hunk 4 above)
Comment (jubrad) [reply src/adapter/src/catalog/open.rs]: ok, this has been simplified quite a bit... I'm just using `is_cluster_size_v2` and keeping much of this code the same as it was.

**Review by jkosh44 — APPROVED** [review]

**Comment by jubrad** [comment]
woot!
image

**Pull request merged**